### DISCUSSION TITLE
[User Manager Service] Auth0 to Mongo relation

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#405] [Q&A] [answered]
### What product are you having troubles with?

Console

### What Console version are you using?

v12.2.1

### Description

Hi, I have a couple of questions about the [User Manager Service](https://docs.mia-platform.eu/docs/runtime_suite/user-manager-service/overview) plugin:
* What is the relation between the MongoDB property `mongoCollection.userGroup` and the Auth0 property `app_metadata.groups`? Are these properties complementary or just alternatives?
I'd like to add the User Manager Service to the actual flow (api-gateway &rarr; authorization-service &rarr; auth0-client) without reworking the entire user groups and permissions flow;

* do I need to import all the users from auth0 to the Mongo collection to see them in the backoffice?

* can I use the `authUserId` field to refer users from other collections and to recognize users in fields like `creatorId` or `updaterId`?

Thanks in advance.

**Comment by theliuk** [comment]
I can answer with full confidence to the second question: yes, users need to be in the CRUD to be shown inside BO.

I leave the other two questions to @nicola88

**Comment by nicola88** [comment, marked as answer]
> What is the relation between the MongoDB property mongoCollection.userGroup and the Auth0 property app_metadata.groups? Are these properties complementary or just alternatives?

The MongoDB field and the Auth0 property have the same value.

> Can I use the authUserId field to refer users from other collections and to recognize users in fields like creatorId or updaterId?

I am not sure I understood your question, but you can configure the UMS to use an existing users collection, as long as it has the schema described in the documentation, and the UMS will basically keep that collection in sync with Auth0.

If you need to do more stuff, like mapping user IDs from other collections using the creatorId or updaterId field, that's entirely up to your application.

**Reply by frengooo** [reply]
Hi @nicola88,
thanks for your reply.

> The MongoDB field and the Auth0 property have the same value.

Can you be more precise? The mongoDB field is updated each time the user logs in? What happen if the values diverge?

> I am not sure I understood your question

I think your reply is enough 😄

**Reply by nicola88** [reply]
The MongoDB field and the Auth0 property are updated automatically when you create or update a user using the UMS API and they are expected not to be modified in any other way. MongoDB is basically considered the source of truth.

**Reply by frengooo** [reply]
Ok, I don't like so much this behaviour because in this way the user can't edit the metadata directly on Auth0.

**Reply by nicola88** [reply]
I understand, but the UMS was designed with the goal of fully manage the user on the CRUD and the identity provider (giving you the option to eventually switch to other identity providers seamlessly), so if you want to change things manually on Auth0 but not in the CRUD maybe the UMS is not the right tool for the job.

**Reply by frengooo** [reply]
And what for other custom user metadata? Can I handle them through the UMS? I didn't understood looking at the documentation. @nicola88

**Reply by nicola88** [reply]
Yes, you can add more fields to the [users CRUD collection](https://docs.mia-platform.eu/docs/runtime_suite/user-manager-service/configuration#user-crud-collection) to store additional information about the users and configure the [`GET /userinfo`](https://docs.mia-platform.eu/docs/runtime_suite/user-manager-service/usage#get-userinfo) to include those fields in the user properties.

EDIT: I fixed the second link which was broken.

**Reply by ffiorentinolabid** [reply]
Hi @nicola88 thanks for clarification on questions that @silversoul93 asked on my behalf.
It's clear that I can add additional information on USER_CRUD_COLLECTION, but how can I manage the relation between them and `app_metadata`  from Auth0. I would like to automatically update on Auth0 other properties: is them managed with the env var `USERINFO_ADDITIONAL_PROPERTIES`?

**Reply by nicola88** [reply]
Those additional user metadata are stored only in the CRUD and when you send a request to the `GET /userinfo` of the UMS, those information will automatically be merged with the userinfo from Auth0. So, if you use the Auth0 client and you configure it to point to the endpoints of the UMS, you should not have to worry where these information are actually stored since basically the UMS enriches the original response from Auth0.

**Reply by ffiorentinolabid** [reply]
On our scenario (migration on backoffice from old CMS, using auth0 on a multitenant application) we always have a pair (brand + group) used to manage with `app_metadata` of Auth0: using UMS we'll have this 'decoupling' not so good.
In any case I would like to try it to understand if it's relevant for us or not: can you help to understand how to have a `brandId` property managed correctly? 
I've added it to USER_CRUD_COLLECTION, but seems unable to retrieve it from `/userinfo`: I've tried to add it on `USERINFO_ADDITIONAL_PROPERTIES`(actually value is `permissions,brandId`).
